### ISSUE TITLE
CVE-2017-12883 and Perl 5.22

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16157]
Migrated from [rt.perl.org#132134](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132134) (status was 'open')

Searchable as RT132134$

**Comment by p5pRT** [comment]
### From z5t1@z5t1.com

To whom it may concern&#8203;:

I have backported your patch for CVE\-2017\-12883 to Perl 5\.22
\(originally\, a patch was released for only Perl 5\.24 and 5\.26\)\. For your
convenience\, I have attached the patch however it can also be found at
http&#8203;://mirror\.cucumberlinux\.com/cucumber/cucumber\-1\.0/source/lang\-base/perl/patches/CVE\-2017\-12883\.patch\.

Sincerely\,

Scott Court
BDFL of Cucumber Linux

**Comment by p5pRT** [comment]
### From z5t1@z5t1.com

<details><summary>CVE-2017-12883.patch</summary>

```diff
--- a/regcomp.c
+++ b/regcomp.c
Fixes CVE-2017-12883 for Perl 5.22
Upstream commit 2be4edede4ae226e2eebd4eff28cedd2041f300f

Note we had to change this patch slightly to get it to work with Perl 5.22.
We did this by taking their official patch URL (https://perl5.git.perl.org/perl.git/blobdiff/f7e5417e7bffba03947b66e4d8622d7c220f2876..40b3cdad3649334585cee8f4630ec9a025e62be6:/regcomp.c)
and changing the first commit to be the Perl 5.22.4 commit (a26666a1317770d8a2228ac3657ba58020c3511f),
which resulted in a URL of https://perl5.git.perl.org/perl.git/blobdiff/a26666a1317770d8a2228ac3657ba58020c3511f..40b3cdad3649334585cee8f4630ec9a025e62be6:/regcomp.c.
We then cherry picked this one change from that diff.
@@ -11303,13 +11303,15 @@
 	}
         sv_catpv(substitute_parse, ")");
 
-	RExC_parse = SvPV(substitute_parse, len);
+	len = SvCUR(substitute_parse);
 
 	/* Don't allow empty number */
 	if (len < (STRLEN) 8) {
             RExC_parse = endbrace;
 	    vFAIL("Invalid hexadecimal number in \\N{U+...}");
 	}
+
+	RExC_parse = SvPV_nolen(substitute_parse);
 	RExC_end = RExC_parse + len;
 
         /* The values are Unicode, and therefore not subject to recoding, but


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed\, 20 Sep 2017 15&#8203;:38&#8203;:31 GMT\, z5t1@&#8203;z5t1\.com wrote&#8203;:
> To whom it may concern&#8203;:
> 
> I have backported your patch for CVE\-2017\-12883 to Perl 5\.22
> \(originally\, a patch was released for only Perl 5\.24 and 5\.26\)\. For
> your
> convenience\, I have attached the patch however it can also be found at
> http&#8203;://mirror\.cucumberlinux\.com/cucumber/cucumber\-1\.0/source/lang\-
> base/perl/patches/CVE\-2017\-12883\.patch\.
> 
> Sincerely\,
> 
> Scott Court
> BDFL of Cucumber Linux

I have made this patch available for smoke testing in the following branch&#8203;:

smoke\-me/jkeenan/132134\-cve\-2017\-12883

Whether it is applied or not is for discussion\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, 20 Sep 2017 17&#8203;:00&#8203;:17 \-0700\, jkeenan wrote&#8203;:
> On Wed\, 20 Sep 2017 15&#8203;:38&#8203;:31 GMT\, z5t1@&#8203;z5t1\.com wrote&#8203;:
> > To whom it may concern&#8203;:
> > 
> > I have backported your patch for CVE\-2017\-12883 to Perl 5\.22
> > \(originally\, a patch was released for only Perl 5\.24 and 5\.26\)\. For
> > your
> > convenience\, I have attached the patch however it can also be found at
> > http&#8203;://mirror\.cucumberlinux\.com/cucumber/cucumber\-1\.0/source/lang\-
> > base/perl/patches/CVE\-2017\-12883\.patch\.
> > 
> > Sincerely\,
> > 
> > Scott Court
> > BDFL of Cucumber Linux
> 
> I have made this patch available for smoke testing in the following branch&#8203;:
> 
> smoke\-me/jkeenan/132134\-cve\-2017\-12883
> 
> Whether it is applied or not is for discussion\.

That looks fine to me\.

Tony

**Comment by khwilliamson** [comment]
What to do about this ticket?

**Comment by demerphq** [comment]
there was no 5.22.5, this patch was made on top of 5.22.4, id say it can be closed as we aren't going to create a 5.22.5 at this point (afaiui)

**Comment by jkeenan** [comment]
> there was no 5.22.5, this patch was made on top of 5.22.4, id say it can be closed as we aren't going to create a 5.22.5 at this point (afaiui)

The smoke-me branch mentioned in this ticket no longer exists, but it was run and reports were collated [here](http://perl.develop-help.com/?b=smoke-me%2Fjkeenan%2F132134-cve-2017-12883).  From that page we can locate this [commit](https://github.com/Perl/perl5/commit/1dc3982143baf4ab9c843ed45fa5b822fd1ecb7c) which holds the OP's "Backport patch for CVE-2017-12883 to Perl 5.22."  My brief eyeballing of the patch suggests that regcomp.c has changed considerably in 5 years, so the patch wouldn't apply to blead.  And, as @demerphq implies, 5.22 is long out-of-support.

@demerphq, I would suggest that you take one look at this commit to see if there's anything of interest in it, then proceed to close it.

**Comment by demerphq** [comment]
FWIW, I have a copy of your smoke-me, as does Richard Leach in his fork.

Anyway, this was a backport of 2be4edede4ae226e2eebd4eff28cedd2041f300f.

I guess we should have released a 5.22.5 but it never happened. From the POV of blead @khwilliamson has already sorted this. He wrote the patch that was backported.

Closing.
